_posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd

```diff
@@ -0,0 +1,757 @@
+
+    cur_layer_input <- x
+    hidden_states <- init_hidden
+    
+    # loop over layers
+    for (i in 1:self$n_layers) {
+      
+      # every layer's hidden state starts from 0 (non-stateful)
+      c(h, c) %<-% hidden_states[[i]]
+      # outputs, of length seq_len, for this layer
+      # equivalently, list of h states for each time step
+      output_sequence <- vector(mode = "list", length = seq_len)
+      
+      # loop over time steps
+      for (t in 1:seq_len) {
+        c(h, c) %<-% self$cell_list[[i]](cur_layer_input[ , t, , , ], list(h, c))
+        # keep track of output (h) for every time step
+        # h has dim (batch_size, hidden_size, height, width)
+        output_sequence[[t]] <- h
+      }
+
+      # stack hs for all time steps over seq_len dimension
+      # stacked_outputs has dim (batch_size, seq_len, hidden_size, height, width)
+      # same as input to forward (x)
+      stacked_outputs <- torch_stack(output_sequence, dim = 2)
+      
+      # pass the list of outputs (hs) to next layer
+      cur_layer_input <- stacked_outputs
+      
+      # keep track of list of outputs or this layer
+      layer_output_list[[i]] <- stacked_outputs
+      # keep track of last state for this layer
+      layer_state_list[[i]] <- list(h, c)
+    }
+ 
+    list(layer_output_list, layer_state_list)
+  }
+    
+)
+
+```
+
+### Calling the `convlstm`
+
+Let's see the input format expected by `convlstm`, and how to access its different outputs.
+
+Here is a suitable input tensor.
+
+```{r}
+# batch_size, seq_len, channels, height, width
+x <- torch_rand(c(2, 4, 3, 16, 16))
+```
+
+First we make use of a single layer.
+
+```{r}
+model <- convlstm(input_dim = 3, hidden_dims = 5, kernel_sizes = 3, n_layers = 1)
+
+c(layer_outputs, layer_last_states) %<-% model(x)
+```
+
+We get back a list of length two, which we immediately split up into the two types of output returned: intermediate outputs from all layers, and final states (of both types) for the last layer.
+
+With just a single layer, `layer_outputs[[1]]`holds all of the layer's intermediate outputs, stacked on dimension two.
+
+```{r}
+dim(layer_outputs[[1]])
+# [1]  2  4  5 16 16
+```
+
+`layer_last_states[[1]]`is a list of tensors, the first of which holds the single layer's final hidden state, and the second, its final cell state.
+
+```{r}
+dim(layer_last_states[[1]][[1]])
+# [1]  2  5 16 16
+dim(layer_last_states[[1]][[2]])
+# [1]  2  5 16 16
+```
+
+For comparison, this is how return values look for a multi-layer architecture.
+
+```{r}
+model <- convlstm(input_dim = 3, hidden_dims = c(5, 5, 1), kernel_sizes = rep(3, 3), n_layers = 3)
+c(layer_outputs, layer_last_states) %<-% model(x)
+
+# for each layer, tensor of size (batch_size, seq_len, hidden_size, height, width)
+dim(layer_outputs[[1]])
+# 2  4  5 16 16
+dim(layer_outputs[[3]])
+# 2  4  1 16 16
+
+# list of 2 tensors for each layer
+length(layer_last_states)
+# 3
```

Comment:
It took me some time to understand the comment and the length here. Maybe print the `str` instead?

Reply:
great idea, done!